sdk/core/core-paging/src/getPagedAsyncIterator.ts

```diff
@@ -18,7 +18,9 @@ export function getPagedAsyncIterator<
 >(
   pagedResult: PagedResult<TPage, TPageSettings, TLink>
 ): PagedAsyncIterableIterator<TElement, TPage, TPageSettings> {
-  const iter = getItemAsyncIterator<TElement, TPage, TLink, TPageSettings>(pagedResult);
+  const iter = getItemAsyncIterator<TElement, TPage, TLink, TPageSettings>(
+    pagedResult as PagedResult<TPage, TPageSettings, TLink, TElement>
```

Comment:
curious why we need this explicit cast here and if we can avoid it

Reply:
The input `pagedResult` doesn't have the fourth generic instantiated so that it remains backward compatible with existing call sites. The cast instantiates that generic to be `TElement` which is the ideal type for `pagedResult` if breaking existing code was not a concern.

Reply:
Do you think having two overloads for `getPagedAsyncIterator` and `getItemAsyncIterator` which take either the existing PagedResult and the other that takes a PagedResult with toElements would help us get rid of the explicit casting?

Reply:
Good point, however, I like to think of overloads or any public-facing additions as things that help our customers and not internal implementations. In this case, I don't believe the customer experience will improve after adding the overload.

Reply:
It has no impact on customers, however, it does help with library maintenance since a) the intent of the types would be very clear b) it protects us against accidentally casting a non-compatible type. With `as unknown as type`, we basically mute the type system which may hide bugs in the future.

If the work of making these overloads is too big, I'd be okay with filing an issue and handling it in a follow-up PR

Reply:
I think I'm slowly understanding some of the issue here. Previously `PagedResult` had no concept of the shape the iterator was providing because were assuming `TPage` was an array and unwrapping the element type from it. But now that we have the case where `TPage` is *not* an array, we have to somehow instead get the typing from the result of `toElements`.

Reply:
To be fair though, do we expect customers to ever call `getPagedAsyncIterator` directly? One would think the dev experience we care most about here would be ourselves?

Reply:
Yes, that's a good summary of the issue encountered in app-config, so that's why we need to create `toElements`. The customers can call `getPagedAsyncIterator`, but they will not be able to use `getItemAsyncIterator` because it is our own internal function.

Reply:
@joheredi and I discussed this further online. I think one thing we could do here is to ditch the new generic all together and have `toElements` return `unknown[]`. The output of `toElements` will have to be cast to `TElement[]` at internal call sites but this provides a simpler starting point for introducing the feature. If strongly typing `toElements` becomes desired, we can re-add the type parameter in a non-breaking way and perhaps with an overload.

@minhanh-phan how does that sound?

Reply:
Sounds good to me!